cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java

```diff
@@ -0,0 +1,109 @@
+import io.cdap.cdap.proto.id.ProgramRunId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.inject.Inject;
+
+/**
+ * A service that periodically scans for program runs that are in Stopping state and
+ * force terminates them if they are running beyond the expected terminateTime.
+ */
+public class ProgramRunStatusMonitorService extends AbstractRetryableScheduledService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ProgramRunStatusMonitorService.class);
+
+  private final Store store;
+  private final ProgramRuntimeService runtimeService;
+  private final int txBatchSize;
+  private final long interval;
+
+  @Inject
+  ProgramRunStatusMonitorService(CConfiguration cConf, Store store, ProgramRuntimeService runtimeService) {
+    this(cConf, store, runtimeService, cConf.getInt(Constants.AppFabric.PROGRAM_TERMINATOR_TX_BATCH_SIZE));
+  }
+
+  @VisibleForTesting
+  ProgramRunStatusMonitorService(CConfiguration cConf, Store store, ProgramRuntimeService runtimeService,
+                                 int txBatchSize) {
+    super(RetryStrategies.fromConfiguration(cConf, Constants.Service.PROGRAM_RUN_STATUS_MONITOR_SERVICE + "."));
+    this.store = store;
+    this.runtimeService = runtimeService;
+    this.txBatchSize = txBatchSize;
+    long specifiedInterval = cConf.getLong(Constants.AppFabric.PROGRAM_TERMINATOR_INTERVAL_MILLIS);
+    if (specifiedInterval <= 0) {
+      LOG.warn("Invalid interval specified for the program terminator {}. Setting it to 5 minutes.",
+               specifiedInterval);
+      specifiedInterval = 30000L;
+    }
+    this.interval = specifiedInterval;
+  }
+
+  @Override
+  protected long runTask() {
+    LOG.debug("Starting ProgramRunStatusMonitorService");
+    terminateStoppingProgramsActiveBeyondTerminateTime();
+    return this.interval;
+  }
+
+  void terminateStoppingProgramsActiveBeyondTerminateTime() {
+    // fetch all runs that are in stopping state that started at most a minute before current time.
+    // Specifying the entire time range should not be worse in performance
+    // than specifying a more restrictive time range because time range is just used as a read-time filter
+    long currentTimeInMillis = System.currentTimeMillis();
+    Map<ProgramRunId, RunRecordDetail> stoppingRuns = store.getRuns(ProgramRunStatus.STOPPING, 0L,
+                                                                    currentTimeInMillis - 60000, txBatchSize, null);
+    for (Map.Entry<ProgramRunId, RunRecordDetail> stoppingRun : stoppingRuns.entrySet()) {
+      RunRecordDetail runRecordDetail = stoppingRun.getValue();
+      Long terminateTime = runRecordDetail.getTerminateTs();
+      // If terminateTime or RuntimeInfo are null, there isn't anything that can be done about it, so we just ignore it
+      if (terminateTime != null && terminateTime <= TimeUnit.MILLISECONDS.toSeconds(currentTimeInMillis)) {
```

Comment:
try to avoid nested ifs when possible. In cases like this, it's more readable to do:

```
for (...) {
  if (terminateTime == null || terminateTime > ...) {
    continue;
  }
  ...
  if (runtimeInfo == null) {
    continue;
  }
  controller = ...;
  if (controller != null) {
    LOG.info(...)
    controller.stop();
  }
}
```

Reply:
Thanks for the clear description :)